src/lib/userStorage/UserProfileStorage.js

```diff
@@ -0,0 +1,556 @@
+        'Fields validation failed',
+        new Error('setProfile failed: Fields validation failed'),
+        { errors, category: ExceptionCategory.Human },
+      )
+
+      throw errors
+    }
+
+    const { avatar } = profile
+
+    if (!!avatar && isValidBase64Image(avatar)) {
+      const cids = await this._resizeAndStoreAvatars(avatar)
+
+      assign(profile, cids)
+    }
+
+    const fieldsToSave = fields.reduce(
+      (acc, currKey) => ({
+        ...acc,
+        [currKey]: {
+          value: profile[currKey],
+          display: profile[currKey],
+          privacy: this.getFieldPrivacy(currKey),
+        },
+      }),
+      {},
+    )
+    if (update) {
+      return this.setProfileFields(fieldsToSave)
+    }
+    return this.setNewProfileFields(fieldsToSave)
+  }
+
+  /**
+   * gets the current unencrypted field values in-memory
+   * @returns
+   */
+  getProfile(): { [key: string]: string } {
+    return Object.keys(this.profile).reduce((acc, currKey) => ({ ...acc, [currKey]: this.profile[currKey].value }), {})
+  }
+
+  /**
+   * updates profile fields in realm
+   * @param {*} fields
+   * @returns
+   */
+  async setProfileFields(fields: { [key: string]: ProfileField }): Promise<any> {
+    const encryptedFields = await this._encryptProfileFields(fields)
+    await this.profiledb.setProfileFields(encryptedFields)
+    this._setLocalProfile({ ...this.profile, ...fields })
+  }
+
+  /**
+   * create new profile with given fields in realm
+   * @param fields
+   * @returns {Promise<void>}
+   * @private
+   */
+  async setNewProfileFields(fields: { [key: string]: ProfileField }): Promise<any> {
+    const encryptedFields = await this._encryptProfileFields(fields)
+    await this.profiledb.setProfile(encryptedFields)
+    this._setLocalProfile({ ...this.profile, ...fields })
+  }
+
+  /**
+   * helper to set display field based on privacy setting
+   * @param field
+   * @param value
+   * @param privacy
+   * @returns {*}
+   * @private
+   */
+  _setDisplayFieldBasedOnPrivacy(field: string, value: string, privacy: string) {
+    let display
+    switch (privacy) {
+      case 'private':
+        display = '******'
+        break
+      case 'masked':
+        display = UserProfileStorage.maskField(field, value)
+
+        //undo invalid masked field
+        if (display === value) {
+          privacy = 'public'
+        }
+        break
+      case 'public':
+        display = value
+        break
+      default:
+        throw new Error('Invalid privacy setting', { privacy })
+    }
+    return display
+  }
+
+  /**
+   * Set profile field with privacy
+   * @param {*} field
+   * @param {*} value
+   * @param {*} privacy
+   * @param {*} onlyPrivacy
+   * @returns
+   */
+  setProfileField(
+    field: string,
+    value: string,
+    privacy: FieldPrivacy = 'public',
+    onlyPrivacy: boolean = false,
+  ): Promise<ACK> {
+    const display = this._setDisplayFieldBasedOnPrivacy(field, value, privacy)
```

Comment:
i think this should be moved to setProfileFields, will solve all the display issues comments

Reply:
setProfileFields require display value